server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/PendingListenersQueue.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.NotMasterException;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class PendingListenersQueue {
+
+    private static final Logger logger = LogManager.getLogger(PendingListenersQueue.class);
+
+    private record PendingListener(long index, ActionListener<Void> listener) {}
+
+    private final ThreadPool threadPool;
+    private final Queue<PendingListener> pendingListeners = new LinkedList<>();
+    private volatile long completedIndex = -1;
+
+    public PendingListenersQueue(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    public void add(long index, ActionListener<Void> listener) {
+        synchronized (pendingListeners) {
+            pendingListeners.add(new PendingListener(index, listener));
+        }
+    }
+
+    public void complete(long index) {
```

Comment:
Can we rename `index` to `convergedIndex` like in the private `executeListeners` method? That would make it clearer to me.